### PR TITLE
[RFC] Introduce `krnl.yield` Operation for `krnl.iterate`

### PR DESCRIPTION
**Pull request by python3kgae**
Introduce the `krnl.yield` operation within the `krnl.iterate` construct.
This operation functions analogously to `affine/scf.yield`,
yielding zero or more SSA values from a `krnl.iterate` operation region and subsequently terminating said region.

With 'krnl.yield', 'krnl.iterate' will be able to have results like affine.for.
krnl.iterate with results will be look like this.
```

%y0, %y1 = krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg2 = 0 to 10, %2#1 -> %arg3 = 0 to 10) iter_args(%arg4 = %cst, %arg5 = %cst) -> (f32, f32) 
    ...
    krnl.yield %val0, %val1 : f32, f32
 }
```
There're 2 iter_args which match the number of results.
The krnl.yield will yield 2 values also.
The init values for iter_args will be at end of the operands for the krnl.yield.

After convert to affine, the above example will be looks like this.
```

%0:2 = affine.for %arg2 = 0 to 10 iter_args(%arg3 = %cst, %arg4 = %cst) -> (f32, f32) {
  %1:2 = affine.for %arg5 = 0 to 10 iter_args(%arg6 = %arg3, %arg7 = %arg4) -> (f32, f32) {
  ...
  affine.yield %val0, %val1: f32, f32
affine.yield %1#0, %1#1 : f32, f32
```
The inner affine.for will use iter_args from outer affine.for as init for iter_args.
The outer affine.for will yield result of inner affine.for.

Implementing this will facilitate the avoidance of `memref.alloca` for storing reduction values.

Signed-off-by: Xiang Li <python3kgae@outlook.com>

### CODE REVIEW

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@python3kgae please observe the DCO signing protocol, see 
https://github.com/onnx/onnx-mlir?search=1#contributing
on how to satisfy the DCO checks.

**Comment by python3kgae** [comment]
> @python3kgae please observe the DCO signing protocol, see https://github.com/onnx/onnx-mlir?search=1#contributing on how to satisfy the DCO checks.

Fixed.

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToKrnl/Math/Softmax.cpp [hunk 1]
```diff
@@ -23,14 +23,19 @@ namespace onnx_mlir {
...
 
+  ValueRange maxInits = ValueRange(negInfinity);
   // Compute the maximum value along axis.
   ValueRange maxLoops = createKrnl.defineLoops(numberOfLoops);
-  createKrnl.iterateIE(maxLoops, maxLoops, Lbs, Ubs,
+  auto maxLoop = createKrnl.iterateIE(maxLoops, maxLoops, Lbs, Ubs, maxInits,
       [&](KrnlBuilder &createKrnl, ValueRange maxIndices) {
+        Block *iterEntryBB = createKrnl.getBuilder().getBlock();
+        // Get last argument for the iterate body.
+        Value iterArg = iterEntryBB->getArguments().back();
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Math/Softmax.cpp:37]: I believe we would access the iteration arguments more frequently. How about creating a krnl op to query an iterArg, say `KrnlGetIterationArgs`, or we can put it into an argument of `createKrnl.iterateIE` similar to `maxIndices` and we can use it inside the body without explicitly getting it.

**Review by tungld — COMMENTED** [review]
It is a worth extension, and thank you for a complete solution! Let's hear comments from other members @AlexandreEichenberger and @chentong319.

File: src/Conversion/ONNXToKrnl/Math/Softmax.cpp (hunk 1 above)
Comment (python3kgae) [reply src/Conversion/ONNXToKrnl/Math/Softmax.cpp]: Updated as argument like maxIndices for the bodybuilder of KrnlIterateOp to access the iteration arguments.
KrnlIterateOp::getRegionIterArgs was already added to access the iteration arguments of KrnlIterateOp.

Comment (tungld) [reply src/Conversion/ONNXToKrnl/Math/Softmax.cpp]: Thank you!

**Review by tungld — APPROVED** [review]
LGTM.

Please wait awhile for comments from @AlexandreEichenberger and @chentong319 if any before landing this.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

**Comment by tungld** [comment]
@jenkins-droid test this please

**Pull request merged**